src/RelayProvider.ts

```diff
@@ -0,0 +1,270 @@
+import { constants, Transaction } from 'ethers';
+import type { Networkish } from "@ethersproject/networks";
+import { JsonRpcProvider, TransactionReceipt } from "@ethersproject/providers";
+import type { ConnectionInfo } from "@ethersproject/web";
+import { getEnvelopingConfig } from './utils';
+import { hexValue } from "@ethersproject/bytes";
+import RelayClient, { RequestConfig } from './RelayClient';
+import log from 'loglevel';
+import { RelayHub__factory } from '@rsksmart/rif-relay-contracts';
+import AccountManager from './AccountManager';
+import type { UserDefinedEnvelopingRequest } from './common/relayRequest.types';
+
+export interface RelayingResult {
+    validUntilTime?: string;
+    transaction: Transaction;
+    receipt: TransactionReceipt;
+}
+
+export default class RelayProvider extends JsonRpcProvider {
+
+    //TODO: remove & when relayTransaction PR is merged and rebased
+    private readonly relayClient!: RelayClient & { relayTransaction: (
+        envelopingRequest: UserDefinedEnvelopingRequest,
+        requestConfig: RequestConfig) => Promise<Transaction> };
+    
+    private readonly jsonRpcProvider: JsonRpcProvider;
+
+    constructor(url?: ConnectionInfo | string, network?: Networkish, relayClient?: RelayClient){
```

Comment:
I'd allow the user to specify either the JsonRpcProvider or the parameters required to build it.

Reply:
got it

Reply:
I meant we should either accept `url` and `network` or a JsonRpcProvider.
If JsonRpcProvider is received, we could extract `url` from `provider.connection.url` and `network` from `provider.network`

Reply:
I have just pushed a change for the constructor, let me know if it is what you meant

Reply:
Actually I was thinking about something like:
```
constructor(
        relayClient = new RelayClient(),
        providerOrProviderParams: Either<JsonRpcProvider,ProviderParams>){
            let url:  ConnectionInfo | string;
            let network: Networkish;

            if (providerOrProviderParams.connection) {
                url = providerOrProviderParams.connection.url;
                network = providerOrProviderParams._network;
            } else {
                url = providerOrProviderParams.url;
                network = providerOrProviderParams.network
            }
            super(url, network);
            this.relayClient = relayClient;
    }
```
WDYT?